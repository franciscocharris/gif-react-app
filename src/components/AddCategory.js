import React, { useState } from 'react';
import PropTypes from 'prop-types';

export const AddCategory = ({setCategories}) => {

    const [inputValue, setInputValue] = useState('');
    const handleInputChange = (e) =>{
        setInputValue( e.target.value  );
    }

    const handleSubmit = (e) => {
        e.preventDefault();
        if(inputValue.trim().length > 0){
            // console.log(`handleSubmit`, inputValue);
            // al llamar a una funcion de un hook, traera los valores si se trata como un callback
            // aqui se esta llenando el array de los gifs que se van a cargar
            setCategories( cat => [inputValue, ...cat] ); 
            setInputValue('');  
        }
    }

    return (
        <form onSubmit={handleSubmit} >
            {/* solo prueba */}
            <p>{ inputValue }</p>
            <input 
                type='text'
                placeholder='Buscar Gifs'
                value={inputValue}
                // el onchange se dispara cada vez que cambien el valor del input
                onChange={ handleInputChange }
            />
        </form>
        
    )
}

AddCategory.propTypes = {
    setCategories: PropTypes.func.isRequired
}