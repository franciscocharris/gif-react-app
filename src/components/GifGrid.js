import React from 'react';
import { useFetchGifs } from '../hooks/useFetchGifs';
// import { getGifs } from '../helpers/getGifs';
import { GifGridItem } from './GifGridItem';

export const GifGrid = ({ category }) => {

// para renombrar con la desestructuracion, solo es dos puntos : y el nuevo nombre
    const { data:images, loading} = useFetchGifs(category);
    
    // el useEffect sirve para ejecutar codigo con una condicion, se le pasa parametros y si es un array vacio
    // entonces significa que se ejecute una sola vez, cuamdo se renderize el componente
    // useEffect( () => {
    //     getGifs( category )
    //         // .then( imgs => setImages(imgs) );
    //         .then( setImages);
    // }, [category]);

    return (
        <>
            <h3 className='animate__animated animate__animated' ><span>{ category }</span></h3>
            {loading && <p className='animate__animated animate__flash'>Loading</p> }
            <div className='card-grid'>
                    {
                        images.map( img => 
                            <GifGridItem 
                                key={img.id} 
                                // se le hace un spread operator a la coleccion de objetos para enviar propiedades de manera
                                // individual
                                {...img}
                            />
                        )
                    }
            </div>
        </>
    )
}
