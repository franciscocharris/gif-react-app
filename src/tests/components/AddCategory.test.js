import React from 'react';
import { shallow } from "enzyme"
import { AddCategory } from "../../components/AddCategory"

describe('pruebas en el componente <AddCategory/>', () => {
    // este jest.fn es para poder hacer varias comprobaciones como si fuera una funcion con ayuda de jest
    const setCategories = jest.fn();
    let wrapper = shallow(<AddCategory setCategories={ setCategories } />);

    beforeEach( () => {
        // esto resetea todos los mooks o valores de funciones o de todo para antes de testear 
        jest.clearAllMocks();
        wrapper = shallow(<AddCategory setCategories={ setCategories } />);
    });
    
    test('debe de mostrarse correctamente', () => {


        expect(wrapper).toMatchSnapshot();
    })

    test('prueba orientada a a solo ver como se puede probar el cambio de un input', () => {
        const input = wrapper.find('input');

        // el segundo parametro, es el evento que se supone el input esta recibiendo
        const value = 'este es un valor';
        input.simulate('change', { target: { value } });
        // console.log(wrapper.find('p').text().trim());
        expect( wrapper.find('p').text().trim() ).toBe(value);
    })
    
    test('no se debe de ejecutar las funciones en el submit', () => {
        // como no se le dio valor al inpuit obviamnete no paso el if y por ende no se ejecuto esa funcion
        //  que en el test de abajo, si se le da un valor
        wrapper.find('form').simulate('submit', { preventDefault(){} });

        expect( setCategories ).not.toHaveBeenCalled();
    });

    test('debe de llamar el setCategories y dejar el inpunt vacio', () => {
        const value = '-valor del input-'
        // se simula el cambio en el input, el second parameter es el evento y lo que se puede hacer con el 
        const input = wrapper.find('input').simulate('change', { target: { value } });
        //  se simula el submit del form y al evento se le da esta funcion 
        wrapper.find('form').simulate('submit', { preventDefault(){} });
        // tuvo que haber sido llamado
        expect(setCategories).toHaveBeenCalled();
        // otras cosas que se pueden hacer
        /////// se esperaba tantas llamadas de esta funcion
        // expect(setCategories).toHaveBeenCalledTimes(2);

        // se esperaba que esta funcion recibiera una funcion 
        //////// expect(setCategories).toHaveBeenCalledWith( expect.any(Function) );

        // ///
        // console.log(wrapper.find('input').props());
        expect(wrapper.find('input').prop('value')).toBe('');
    })
    // quede en el video nueve seccion 8
    
    
})
