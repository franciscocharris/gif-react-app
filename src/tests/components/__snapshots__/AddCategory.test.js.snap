// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`pruebas en el componente <AddCategory/> debe de mostrarse correctamente 1`] = `
<form
  onSubmit={[Function]}
>
  <p />
  <input
    onChange={[Function]}
    placeholder="Buscar Gifs"
    type="text"
    value=""
  />
</form>
`;
